### PR TITLE
Make Documents mutable

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This is a port of https://github.com/firebase/firebase-android-sdk/pull/2383 and https://github.com/firebase/firebase-android-sdk/pull/2468. The goal is to reduce the number of object creations, so that we can have clear object ownership in the C++ clients. With this change, the documents are owned either by the RemoteDocumentCache or by the View that contains them. Mutations mutate these existing documents and do not create copies.

The existing document hierarchy is removed and replaced by a single MutableDocument type that allows documents to transition between four states:
- INVALID, which is a document we only created so we can later apply a mutation on it. This allows documents to be created by SetMutations
- FOUND, corresponds to the old Document
- NO_DOCUMENT, corresponds to the old NoDocument
- UNKNOWN_DOCUMENT, corresponds to the old UnknownDocument

I kept Document, but it is now an interface that makes sure that we no longer modify the documents once we store them in View. MutableDocuments are produces by the RemoteDocumentCache and via RemoteEvents. We then apply all mutations and from then onwards treat them as Documents.

ObjectValue and ObjectValueBuilder are also merged to reduce immutable object creation.

There should be no functional changes. https://github.com/firebase/firebase-android-sdk/pull/2383 contains a bunch of discussions that might help with this review.

### CODE REVIEW

File: packages/firestore/src/core/view.ts [hunk 1]
```diff
@@ -154,90 +154,78 @@ export class View {
...
-        if (newDoc) {
-          debugAssert(
-            key.isEqual(newDoc.key),
-            'Mismatching keys found in document changes: ' +
-              key +
-              ' != ' +
-              newDoc.key
-          );
-          newDoc = queryMatches(this.query, newDoc) ? newDoc : null;
-        }
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/view.ts:170]: This should be the only actual change here.

File: packages/firestore/test/unit/model/object_value_builder.test.ts [hunk 2]
```diff
@@ -17,78 +17,75 @@
...
-} from '../../../src/model/object_value';
+import { ObjectValue } from '../../../src/model/object_value';
 import { field, wrap, wrapObject } from '../../util/helpers';
 
-describe('ObjectValueBuilder', () => {
-  it('supports empty builders', () => {
-    const builder = new ObjectValueBuilder();
-    const object = builder.build();
-    expect(object.isEqual(ObjectValue.empty())).to.be.true;
+describe('MutableObjectValue', () => {
```
Comment (schmidt-sebastian) [inline packages/firestore/test/unit/model/object_value_builder.test.ts:23]: I would like to rename this file to `mutable_object_value.test.ts` but GitHub then sees it as a new file, so I will do this after the first review.

**Review by thebrianchen — COMMENTED** [review]
Mostly documentation nits with a few questions. Thanks for taking the time to go over this with me!

File: packages/firestore/src/model/mutation.ts [hunk 3]
```diff
@@ -427,54 +393,43 @@ export class SetMutation extends Mutation {
...
 function applySetMutationToLocalView(
   mutation: SetMutation,
-  maybeDoc: MaybeDocument | null,
+  document: MutableDocument,
   localWriteTime: Timestamp
-): MaybeDocument | null {
-  if (!preconditionIsValidForDocument(mutation.precondition, maybeDoc)) {
-    return maybeDoc;
+): void {
+  if (!preconditionIsValidForDocument(mutation.precondition, document)) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation.ts:413]: I removed this and replaced it with an assert, but it turns out I forgot about `CollectionReference.add()`. Add verifies that the newly added document doesn't yet exist.

File: packages/firestore/test/util/helpers.ts [hunk 4]
```diff
@@ -708,7 +713,7 @@ export function documentSet(...args: unknown[]): DocumentSet {
   }
   for (const doc of args) {
     debugAssert(
-      doc instanceof Document,
+      doc instanceof MutableDocument,
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/test/util/helpers.ts:716]: Interfaces can't be used in instanceof checks.

File: packages/firestore/test/util/helpers.ts [hunk 5]
```diff
@@ -664,15 +669,15 @@ export function mapAsArray<K, V>(
  */
 export function documentUpdates(
   ...docsOrKeys: Array<Document | DocumentKey>
-): MaybeDocumentMap {
-  let changes = maybeDocumentMap();
+): DocumentMap {
+  let changes = documentMap();
   for (const docOrKey of docsOrKeys) {
-    if (docOrKey instanceof Document) {
+    if (docOrKey instanceof MutableDocument) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/test/util/helpers.ts:675]: Interfaces can't be used in instanceof checks.

File: packages/firestore/src/model/mutation.ts [hunk 6]
```diff
@@ -222,13 +218,13 @@ export abstract class Mutation {
...
- * expected state (e.g. it is null or outdated), an `UnknownDocument` can be
- * returned.
+ * Applies this mutation to the given Document for the purposes of computing a
+ * new remote document. If the input document doesn't match the expected state
+ * (e.g. it is invalid or outdated), the document state may transition to
+ * unknown.
  *
  * @param mutation - The mutation to apply.
- * @param maybeDoc - The document to mutate. The input document can be null if
+ * @param document - The document to mutate. The input document can be null if
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation.ts]: Done

File: packages/firestore/src/model/mutation.ts [hunk 7]
```diff
@@ -237,38 +233,30 @@ export abstract class Mutation {
...
-    );
+    applyDeleteMutationToRemoteDocument(mutation, document, mutationResult);
   }
 }
 
 /**
- * Applies this mutation to the given MaybeDocument or null for the purposes
- * of computing the new local view of a document. Both the input and returned
- * documents can be null.
+ * Applies this mutation to the given document or null for the purposes of
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation.ts:249]: Done

File: packages/firestore/src/model/mutation_batch.ts [hunk 8]
```diff
@@ -67,23 +67,15 @@ export class MutationBatch {
    * to create a new remote document
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation_batch.ts:67]: Done

File: packages/firestore/src/model/mutation_batch.ts [hunk 9]
```diff
@@ -67,23 +67,15 @@ export class MutationBatch {
...
    * @param batchResult - The result of applying the MutationBatch to the
    * backend.
    */
   applyToRemoteDocument(
     docKey: DocumentKey,
-    maybeDoc: MaybeDocument | null,
+    document: MutableDocument,
     batchResult: MutationBatchResult
-  ): MaybeDocument | null {
-    if (maybeDoc) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation_batch.ts:79]: `docKey` wasn't used anymore, but I forgot to remove it.

**Review by schmidt-sebastian — COMMENTED** [review]
Thanks for the thorough review! Someone deserves a vacation!

**Review by thebrianchen — APPROVED** [review]
lgtm!

File: packages/firestore/src/lite/reference_impl.ts [hunk 10]
```diff
@@ -134,7 +133,7 @@ export function getDoc<T>(
         reference.firestore,
         userDataWriter,
         reference._key,
-        maybeDocument instanceof Document ? maybeDocument : null,
+        maybeDocument.isFoundDocument() ? maybeDocument : null,
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/lite/reference_impl.ts]: Done

File: packages/firestore/src/local/local_serializer.ts [hunk 11]
```diff
@@ -63,6 +58,7 @@ import {
   DbUnknownDocument
 } from './indexeddb_schema';
 import { TargetData, TargetPurpose } from './target_data';
+import { Timestamp } from '../exp/timestamp';
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/local/local_serializer.ts]: I think this was already fixed.

File: packages/firestore/src/model/document.ts [hunk 12]
```diff
@@ -16,168 +16,339 @@
...
 
 import { DocumentKey } from './document_key';
 import { ObjectValue } from './object_value';
 import { FieldPath } from './path';
 import { valueCompare } from './values';
 
-export interface DocumentOptions {
-  hasLocalMutations?: boolean;
-  hasCommittedMutations?: boolean;
+const enum DocumentType {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/document.ts:30]: Done

File: packages/firestore/src/model/mutation.ts [hunk 13]
```diff
@@ -222,76 +215,65 @@ export abstract class Mutation {
 }
 
 /**
- * Applies this mutation to the given MaybeDocument or null for the purposes
- * of computing a new remote document. If the input document doesn't match the
- * expected state (e.g. it is null or outdated), an `UnknownDocument` can be
- * returned.
+ * Applies this mutation to the given document for the purposes of computing a
+ * new remote document. If the input document doesn't match the expected state
+ * (e.g. it is invalid or outdated), the document state may transition to
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation.ts]: Done

**Pull request merged**